### PR TITLE
Add group api

### PR DESCRIPTION
**Pull request by snackwangpetter**

### CODE REVIEW

File: src/android/CarrierPlugin.java [hunk 1]
```diff
@@ -293,10 +342,15 @@ private void createObject(JSONArray args, CallbackContext callbackContext) throw
...
           PluginCarrierHandler carrierHandler = PluginCarrierHandler.createInstance(dir, config, mCarrierCallbackContext, this);
+
           if (carrierHandler != null) {
+              pluginGroupHandler = PluginGroupHandler.createInstance(carrierHandler.mCode,mGroupCallbackContext);
+//              pluginGroupHandler.setListener(mGroupCallbackContext);
+
               mCarrierMap.put(carrierHandler.mCode, carrierHandler);
               JSONObject r = new JSONObject();
               r.put("id", carrierHandler.mCode);
+              Context context = this.cordova.getActivity().getApplicationContext();
```
Comment (stiartsly) [inline src/android/CarrierPlugin.java]: variable "context" not used in the context of this function.

File: src/android/CarrierPlugin.java [hunk 2]
```diff
@@ -568,11 +622,12 @@ private void sendFriendMessage(JSONArray args, CallbackContext callbackContext)
           String message = args.getString(2);
           PluginCarrierHandler carrierHandler = mCarrierMap.get(id);
           if (carrierHandler != null) {
-              carrierHandler.mCarrier.sendFriendMessage(to, message);
-              callbackContext.success("success!");
+              boolean isOffline = carrierHandler.mCarrier.sendFriendMessage(to, message);
+              //TODO
+              callbackContext.success(SUCCESS);
```
Comment (stiartsly) [inline src/android/CarrierPlugin.java]: "isOffline" should be forwarded to upper javascript layer through "callbackContext.success(xxx)"

File: src/android/CarrierPlugin.java [hunk 3]
```diff
@@ -953,7 +1010,282 @@ private void closePortForwarding(JSONArray args, CallbackContext callbackContext
...
+          PluginCarrierHandler carrierHandler = mCarrierMap.get(id);
+
+          if (carrierHandler != null) {
+              Group group = carrierHandler.mCarrier.newGroup(pluginGroupHandler);
+
+              String groupId = addGroupMap(group);
+
+              JSONObject jsonObject = new JSONObject();
+              jsonObject.put("groupId",groupId);
+              jsonObject.put("groupTitle",group.getTitle());
```
Comment (stiartsly) [inline src/android/CarrierPlugin.java]: where is the source of group.getTitle() ???

File: src/android/CarrierPlugin.java [hunk 4]
```diff
@@ -953,7 +1010,282 @@ private void closePortForwarding(JSONArray args, CallbackContext callbackContext
...
+
+              Toast.makeText(this.cordova.getActivity().getApplicationContext(),"createGroup id="+id +"\n"+"result = "+group.getTitle(),Toast.LENGTH_LONG).show();
```
Comment (stiartsly) [inline src/android/CarrierPlugin.java]: Toast.makeText(...) is a  kind of debug statement, please remove it .

File: src/android/CarrierPlugin.java (hunk 1 above)
Comment (snackwangpetter) [reply src/android/CarrierPlugin.java]: Debug code , i’ll remove it

File: src/android/CarrierPlugin.java (hunk 2 above)
Comment (snackwangpetter) [reply src/android/CarrierPlugin.java]: Fixed

File: src/android/CarrierPlugin.java (hunk 3 above)
Comment (snackwangpetter) [reply src/android/CarrierPlugin.java]: Set default group title "Untitled"

File: src/android/CarrierPlugin.java (hunk 4 above)
Comment (snackwangpetter) [reply src/android/CarrierPlugin.java]: Debug code , i’ll remove it

File: src/android/PluginGroupHandler.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        }
+    }
+
+    @Override
+    public void onGroupMessage(Group group, String from, byte[] message) {
+        JSONObject r = new JSONObject();
+        String messageData = new String(message, StandardCharsets.UTF_8);
+        try {
+            r.put("name", "onGroupMessage");
+            r.put("groupTitle", group.getTitle());
```
Comment (stiartsly) [inline src/android/PluginGroupHandler.java]: Do we need group title here to be reported to upper javascript layer?

File: src/android/PluginGroupHandler.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void onPeerName(Group group, String peerId, String peerName) {
+        JSONObject r = new JSONObject();
+        try {
+            r.put("name", "onPeerName");
+            r.put("groupTitle", group.getTitle());
```
Comment (stiartsly) [inline src/android/PluginGroupHandler.java]: Do we need group title here to be reported to upper javascript layer?

File: src/android/PluginGroupHandler.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void onPeerListChanged(Group group) {
+        JSONObject r = new JSONObject();
+        try {
+            r.put("name", "onPeerListChanged");
+            r.put("groupTitle", group.getTitle());
```
Comment (stiartsly) [inline src/android/PluginGroupHandler.java]: Do we need group title here to be reported to upper javascript layer?

File: src/android/PluginGroupHandler.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    private void sendEvent(JSONObject info) throws JSONException {
+        info.put("id", mCode);
+        if (mCallbackContext != null) {
+         PluginResult result = new PluginResult(PluginResult.Status.OK, info);
+         result.setKeepCallback(true);
+         mCallbackContext.sendPluginResult(result);
```
Comment (stiartsly) [inline src/android/PluginGroupHandler.java]: keep indent with 4 spaces.

File: www/carrier.js [hunk 9]
```diff
@@ -1074,6 +1151,137 @@
            this.process(onSuccess, onError, "replyFriendInvite", [this.objId, to, status, reason, data]);
        },
 
+       /**
+        * Create a new group request.
+        *
+        * This function will create a new group.
+        *
+        * @param {Function} onSuccess  The function to call when success.The param is a json string , contains groupId and groupTitle,like this{"groupId":GROUP_ID,"groupTitle":GROUP_TITILE}.
```
Comment (stiartsly) [inline www/carrier.js]: Native (android/ios) --> groupId

**Pull request merged**